File: app/src/main/java/com/wipro/android/proficiencyexercise/view/list/ListFragmentViewModel.java
package com.wipro.android.proficiencyexercise.view.list;

import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import com.google.gson.Gson;
import com.wipro.android.proficiencyexercise.data.remote.response.CanadaList;
import com.wipro.android.proficiencyexercise.domain.intractor.CanadaListUseCase;
import com.wipro.android.proficiencyexercise.utils.LogUtil;

import javax.inject.Inject;

import io.reactivex.observers.DisposableObserver;

public class ListFragmentViewModel extends ViewModel {

    private final String TAG = ListFragmentViewModel.class.getSimpleName();
    public final MutableLiveData<Object> apiResponse = new MutableLiveData<>();
    public final MutableLiveData<Boolean> loaderData = new MutableLiveData<>();
    private CanadaListUseCase useCase;

    @Inject
    public ListFragmentViewModel(CanadaListUseCase canadaListUseCase) {
        this.useCase = canadaListUseCase;
    }

    @Override
    protected void onCleared() {
        loaderData.setValue(false);
    }

    public void apiCall() {
        loaderData.setValue(true);
        useCase.execute(new DisposableObserver<CanadaList>() {
            @Override
            public void onNext(CanadaList canadaList) {
                loaderData.setValue(false);
                LogUtil.INSTANCE.d(TAG, new Gson().toJson(canadaList));
                apiResponse.setValue(canadaList);
            }

            @Override
            public void onError(Throwable e) {
                LogUtil.INSTANCE.d(TAG, e.getMessage());
                apiResponse.setValue(e);
                loaderData.setValue(false);
            }

            @Override
            public void onComplete() {
                LogUtil.INSTANCE.d(TAG, " Response Complete");
            }
        });
    }
}

File: app/src/main/java/com/wipro/android/proficiencyexercise/view/list/ListFragmentViewModel.kt
package com.wipro.android.proficiencyexercise.view.list

import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.google.gson.Gson
import com.wipro.android.proficiencyexercise.data.remote.response.CanadaList
import com.wipro.android.proficiencyexercise.domain.intractor.CanadaListUseCase
import com.wipro.android.proficiencyexercise.utils.LogUtil
import io.reactivex.observers.DisposableObserver
import javax.inject.Inject

class ListFragmentViewModel @Inject
constructor(private val useCase: CanadaListUseCase) : ViewModel() {

    val tag = "ListFragmentViewModel"
    val apiResponse = MutableLiveData<CanadaList>()
    val loaderData = MutableLiveData<Boolean>()

    override fun onCleared() {
        loaderData.value = false
    }

    fun apiCall() {
        loaderData.value = true
        useCase.execute(object : DisposableObserver<CanadaList>() {
            override fun onNext(canadaList: CanadaList) {
                loaderData.value = false
                LogUtil.d(tag, Gson().toJson(canadaList))
                apiResponse.value = canadaList
            }

            override fun onError(e: Throwable) {
                LogUtil.d(tag, e.message!!)
                loaderData.value = false
            }

            override fun onComplete() {
                LogUtil.d(tag, " Response Complete")
            }
        })
    }
}
